waiter/src/waiter/interstitial.clj

```diff
@@ -0,0 +1,238 @@
+;;
+;;       Copyright (c) 2017 Two Sigma Investments, LP.
+;;       All Rights Reserved
+;;
+;;       THIS IS UNPUBLISHED PROPRIETARY SOURCE CODE OF
+;;       Two Sigma Investments, LP.
+;;
+;;       The copyright notice above does not evidence any
+;;       actual or intended publication of such source code.
+;;
+(ns waiter.interstitial
+  (:require [clj-time.core :as t]
+            [clojure.core.async :as async]
+            [clojure.java.io :as io]
+            [clojure.set :as set]
+            [clojure.string :as str]
+            [clojure.tools.logging :as log]
+            [comb.template :as template]
+            [metrics.counters :as counters]
+            [plumbing.core :as pc]
+            [waiter.async-utils :as au]
+            [waiter.metrics :as metrics]
+            [waiter.utils :as utils]))
+
+(defn- interstitial-state-initialized?
+  "Returns true if the interstitial-state-atom has been updated by a scheduler state update."
+  [interstitial-state-atom]
+  (:initialized? @interstitial-state-atom))
+
+(defn- service-id->interstitial-promise
+  "Returns the promise mapped against a service-id.
+   It will return nil if the service is not currently mapped to a promise."
+  [interstitial-state-atom service-id]
+  (get-in @interstitial-state-atom [:service-id->interstitial-promise service-id]))
+
+(defn- resolve-promise!
+  "Resolves a promise to the specified value.
+   It must be called inside a critical section to avoid data races."
+  [interstitial-promise interstitial-resolution]
+  (deliver interstitial-promise interstitial-resolution)
+  (when (= interstitial-resolution (deref interstitial-promise))
+    (counters/inc! (metrics/waiter-counter "interstitial" "promise" "resolved"))
```

Comment:
How can the promise be resolved more than once?  I noticed the picture you posted in the PR description has a value of `3` for this.

Reply:
Ignore.  I realized later that there aren't independent counters per service.